cpu/stm32_common/periph/flashpage.c

```diff
@@ -27,60 +30,150 @@
 
 #include "periph/flashpage.h"
 
+#if defined(CPU_FAM_STM32L0) | defined(CPU_FAM_STM32L1)
+/* Data EEPROM and control register unlock keys */
+#define FLASH_KEY1             ((uint32_t)0x89ABCDEF)
+#define FLASH_KEY2             ((uint32_t)0x02030405)
+/* Program memory unlock keys */
+#define FLASH_PRGKEY1          ((uint32_t)0x8C9DAEBF)
+#define FLASH_PRGKEY2          ((uint32_t)0x13141516)
+#define CNTRL_REG              (FLASH->PECR)
+#define CNTRL_REG_LOCK         (FLASH_PECR_PELOCK)
+#define KEY_REG                (FLASH->PEKEYR)
+#define FLASH_CR_PER           (FLASH_PECR_ERASE | FLASH_PECR_PROG)
+#define FLASH_CR_PG            (FLASH_PECR_FPRG | FLASH_PECR_PROG)
+#define FLASHPAGE_DIV          (4U) /* write 4 bytes in one go */
+#else
+#define CNTRL_REG              (FLASH->CR)
+#define CNTRL_REG_LOCK         (FLASH_CR_LOCK)
+#define KEY_REG                (FLASH->KEYR)
+#define FLASHPAGE_DIV          (2U)
+#endif
+
+static void _unlock(void)
+{
+    DEBUG("[flashpage] unlocking the flash module\n");
+    if (CNTRL_REG & CNTRL_REG_LOCK) {
+        KEY_REG = FLASH_KEY1;
+        KEY_REG = FLASH_KEY2;
+    }
+
+#if defined(CPU_FAM_STM32L0) || defined(CPU_FAM_STM32L1)
+    DEBUG("[flashpage] unlocking the flash program memory\n");
+    if (!(CNTRL_REG & CNTRL_REG_LOCK)) {
+        if (CNTRL_REG & FLASH_PECR_PRGLOCK) {
+            DEBUG("[flashpage] setting the program memory unlock keys\n");
+            FLASH->PRGKEYR = FLASH_PRGKEY1;
+            FLASH->PRGKEYR = FLASH_PRGKEY2;
+        }
+    }
+#endif
+}
+
+static void _lock(void)
+{
+    DEBUG("[flashpage] locking the flash module\n");
+    CNTRL_REG |= CNTRL_REG_LOCK;
+}
+
+void flashpage_write_raw(void *target_addr, void *data, size_t len)
+{
+    /* The actual minimal block size for writing is 16B, thus we
```

Comment:
is it really 16B ?

Reply:
During my tests, it was actually possible to write even 4B, but I left 16B as the smallest.